### PR TITLE
Fix deferred operation error

### PR DESCRIPTION
**Pull request by mizhen**
This pull request includes the following changes on both capture and replay side:
if vkCreateRayTracingPipelinesKHR is a deferred operation,the lifetime of
the input parameters are extended to cover related
vkDeferredOperationJoinKHR and vkGetDeferredOperationResultKHR until the
return value show the deferred operation be finished. The post handling of
vkCreateRayTracingPipelinesKHR now is also placed after the deferred
operation be finished.

### CODE REVIEW

File: framework/encode/deferred_operation.cpp
Comment (bradgrantham-lunarg) [inline framework/encode/deferred_operation.cpp:46]: I recommend you make this a bare pointer to be consistent with other singleton instances like `static VulkanCaptureManager* instance_;` in `framework/encode/vulkan_capture_manager.h`.  Another possibility is to make it a `shared_ptr`, so no one has to guess about what happens after this reference is returned.

File: framework/encode/vulkan_capture_manager.cpp
Comment (bradgrantham-lunarg) [inline framework/encode/vulkan_capture_manager.cpp:1071]: I think the use of an rvalue-reference here is confusing.  Is there a reason this can't just be a `shared_ptr`?

**Review by bradgrantham-lunarg — CHANGES_REQUESTED** [review]

File: framework/decode/vulkan_replay_consumer_base.cpp
Comment (rurra-amd) [reply framework/decode/vulkan_replay_consumer_base.cpp:5631]: Same thing below

File: framework/encode/capture_manager.h
Comment (rurra-amd) [reply framework/encode/capture_manager.h:55]: Did Ming made them public to access from DeferredOperationCreateRayTracingPipelines ?

File: framework/encode/deferred_operation.h
Comment (rurra-amd) [inline framework/encode/deferred_operation.h:38]: deferred

File: framework/encode/capture_manager.h
Comment (bradgrantham-lunarg) [reply framework/encode/capture_manager.h:55]: Yes!  Thanks for pointing that out.

File: framework/generated/generated_vulkan_api_call_encoders.cpp
Comment (rurra-amd) [reply framework/generated/generated_vulkan_api_call_encoders.cpp:4]: Earlier today I explained to him how to run the codegen.

File: framework/encode/deferred_operation.cpp
Comment (rurra-amd) [inline framework/encode/deferred_operation.cpp:38]: Curious - @mizhen are you using all these includes? If not then please reduce

File: framework/decode/deferred_operation_info_create_ray_tracing_pipelines.h
Comment (mizhen) [reply framework/decode/deferred_operation_info_create_ray_tracing_pipelines.h:69]: It's a a guarantee of lifetime. For vkCreateRayTracingPipelinesKHR, if the call is not a deferred operation, unwrapped pCreateInfos or other data will be used locally, allocated in
vkCreateRayTracingPipelinesKHR API handler and will be freed in same handler. But if the call is a deferred operation, all those data struct be just allocated in this API handler and freed at other API handler which might be in other thread when detected the deferred operation has ended.  The ownership transfer is used here.  In short, the ownership transfer will happen ( by methods like the following SetModifiedCreateInfos) when these data be allocated and freed in different API handler or the deferred operation cannot be finished in the call.

Comment (bradgrantham-lunarg) [reply framework/decode/deferred_operation_info_create_ray_tracing_pipelines.h:69]: I won't approve this PR unless you change your use of unique_ptr like I suggested above.

Comment (bradgrantham-lunarg) [reply framework/decode/deferred_operation_info_create_ray_tracing_pipelines.h:69]: I think this is harder to read than `shared_ptr` and is unsafe, and is unnecessarily confusing.  First, your code doesn't do something unsafe, but having a bare reference returned to an internal struct member means another engineer might do something unsafe.  The reference's lifetime is not tied to the object's lifetime.  Code in a totally different place is altering the contents of the member through a reference.  I've provided an example of how that could go badly at https://godbolt.org/z/Khj1eq5Ko.

Any of my suggestions above would be clearer and allow a new engineer in the code to better understand and use  this object more easily.

Comment (bradgrantham-lunarg) [reply framework/decode/deferred_operation_info_create_ray_tracing_pipelines.h:69]: I've searched all the `gfxreconstruct` code.  Unless my search was wrong, your PR will be the only addition that uses these idioms, so it will also be confusing for anyone else to use for that reason.

Comment (mizhen) [reply framework/decode/deferred_operation_info_create_ray_tracing_pipelines.h:69]: ok, will change to shared_ptr.

File: framework/encode/deferred_operation.cpp
Comment (mizhen) [reply framework/encode/deferred_operation.cpp:46]: ok, will change to shared_ptr.

File: framework/encode/vulkan_capture_manager.cpp
Comment (mizhen) [reply framework/encode/vulkan_capture_manager.cpp:1071]: ok, will change to shared_ptr.

**Comment by bradgrantham-lunarg** [comment]
@mizhen Thank you for this PR!  I apologize for how long it's taken to resolve.  As this functionality has been merged in another PR, I'm closing this PR.

**Pull request closed without merging**
